### PR TITLE
Some fixes for LDAP

### LINKED ISSUE #65762: Tests `test_ldap_external_user_directory` occasionally flaky
Integration tests `test_ldap_external_user_directory` are [flaky](https://play.clickhouse.com/play?user=play#c2VsZWN0IHB1bGxfcmVxdWVzdF9udW1iZXIgYXMgcHIsIGNoZWNrX3N0YXJ0X3RpbWUsIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgdGVzdF9zdGF0dXMgYXMgc3RhdHVzLCByZXBvcnRfdXJsCmZyb20gY2hlY2tzCndoZXJlIHRvRGF0ZShjaGVja19zdGFydF90aW1lKSA+ICcyMDIzLTA3LTI4JyBhbmQgdGVzdF9uYW1lIGxpa2UgJyV0ZXN0X2xkYXBfZXh0ZXJuYWxfdXNlcl9kaXJlY3RvcnklJyBhbmQgc3RhdHVzIGluICgnRkFJTCcsICdFUlJPUicsICdGTEFLWScpCm9yZGVyIGJ5IGNoZWNrX3N0YXJ0X3RpbWUgZGVzYw==)

Example [report](https://s3.amazonaws.com/clickhouse-test-reports/63206/bf93c3807014f4323960afb953565a48e3c26109/integration_tests__aarch64__[4_6].html)

It is interesting that it has not failed for months, and now it only fails in aarch64.

```
=================================== FAILURES ===================================
___________________________ test_authentication_pass ___________________________
[gw3] linux -- Python 3.10.12 /usr/bin/python3

    def test_authentication_pass():
>       assert instance.query(
            "select currentUser()", user="janedoe", password="qwerty"
        ) == TSV([["janedoe"]])

test_ldap_external_user_directory/test.py:52: 

.........................................................................
.........................................................................

E           helpers.client.QueryRuntimeException: Client failed! Return code: 4, stderr: Code: 516. DB::Exception: Received from 172.16.9.3:9000. DB::Exception: janedoe: Authentication failed: password is incorrect, or there is no user with such name.. Stack trace:
E           
E           0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000c343e08
E           1. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x0000000007b9f21c
E           2. DB::AccessControl::authenticate(DB::Credentials const&, Poco::Net::IPAddress const&, String const&) const @ 0x000000000f10cce4
E           3. DB::Session::authenticate(DB::Credentials const&, Poco::Net::SocketAddress const&) @ 0x000000001024ff20
E           4. DB::TCPHandler::receiveHello() @ 0x00000000112643b4
E           5. DB::TCPHandler::runImpl() @ 0x000000001125ab4c
E           6. DB::TCPHandler::run() @ 0x0000000011273048
E           7. Poco::Net::TCPServerConnection::start() @ 0x00000000135ae698
E           8. Poco::Net::TCPServerDispatcher::run() @ 0x00000000135aeb74
E           9. Poco::PooledThread::run() @ 0x0000000013579574
E           10. Poco::ThreadImpl::runnableEntry(void*) @ 0x0000000013577950
E           11. ? @ 0x000000000007d5c8
E           12. ? @ 0x00000000000e5edc
E           . (AUTHENTICATION_FAILED)

helpers/client.py:239: QueryRuntimeException
------------------------------ Captured log call -------------------------------
2024-06-26 04:06:01 [ 517 ] DEBUG : Executing query select currentUser() on instance (cluster.py:3449, query)
```

### LINKED ISSUE #69555: Crash with ldap on 24.8.4.13 (current latest)
Trace:
```
2024.09.12 12:31:23.295053 [ 14406 ] {b60cfd44-8c83-497e-ba38-1c479869cd82} <Fatal> GWPAsan: *** GWP-ASan detected a memory error ***
2024.09.12 12:31:23.295358 [ 14406 ] {b60cfd44-8c83-497e-ba38-1c479869cd82} <Fatal> GWPAsan: Use After Free (warning: buffer overflow/underflow detected on a free()'d allocation. This either means you have a buffer-overflow and a use-after-free at the same time, or you have a long-lived use-after-free bug where the allocation/deallocation metadata below has already been overwritten and is likely bogus) at 0x281473304510535 (71 bytes to the right of a 72-byte allocation at 0x281473304510464) by thread 14406 here:
2024.09.12 12:31:23.295626 [ 14406 ] {b60cfd44-8c83-497e-ba38-1c479869cd82} <Fatal> GWPAsan: 0x281473304510535 was deallocated by thread 14406 here:
2024.09.12 12:31:23.295815 [ 14406 ] {b60cfd44-8c83-497e-ba38-1c479869cd82} <Fatal> GWPAsan: <Empty trace>
2024.09.12 12:31:23.295988 [ 14406 ] {b60cfd44-8c83-497e-ba38-1c479869cd82} <Fatal> GWPAsan: 0x281473304510535 was allocated by thread 14406 here:
2024.09.12 12:31:23.296159 [ 14406 ] {b60cfd44-8c83-497e-ba38-1c479869cd82} <Fatal> GWPAsan: <Empty trace>
2024.09.12 12:31:23.296370 [ 14406 ] {b60cfd44-8c83-497e-ba38-1c479869cd82} <Fatal> GWPAsan: *** End GWP-ASan report ***
2024.09.12 12:31:23.297268 [ 15136 ] {} <Fatal> BaseDaemon: ########## Short fault info ############
2024.09.12 12:31:23.297588 [ 15136 ] {} <Fatal> BaseDaemon: (version 24.8.4.13 (official build), build id: 7A2B16FBACA77EB7D47F452AE6FCC7216C28FDE1, git hash: 53195bc189b587afaabc52c4ca765aa54beea84f) (from thread 14406) Received signal 11
2024.09.12 12:31:23.297904 [ 15136 ] {} <Fatal> BaseDaemon: Signal description: Segmentation fault
2024.09.12 12:31:23.298128 [ 15136 ] {} <Fatal> BaseDaemon: Address: 0xffff9c544047. Access: . Attempted access has violated the permissions assigned to the memory area.
2024.09.12 12:31:23.298311 [ 15136 ] {} <Fatal> BaseDaemon: Stack trace: 0x000000000c558228 0x0000ffff9dffb7dc 0x000000000ee226a1 0x000000000ee1dbac 0x000000000ee21b78 0x000000000eda3fa8 0x000000000ed95bc4 0x000000000edff1e4 0x000000000ff172c8 0x000000000ff16c98 0x000000000fd8a6a0 0x000000000fd86a08 0x0000000010caef0c 0x0000000010cc5588 0x000000001344be98 0x000000001344c374 0x00000000134173b4 0x00000000134157f0 0x0000ffff9df89624 0x0000ffff9dee062c
2024.09.12 12:31:23.298477 [ 15136 ] {} <Fatal> BaseDaemon: ########################################
2024.09.12 12:31:23.298812 [ 15136 ] {} <Fatal> BaseDaemon: (version 24.8.4.13 (official build), build id: 7A2B16FBACA77EB7D47F452AE6FCC7216C28FDE1, git hash: 53195bc189b587afaabc52c4ca765aa54beea84f) (from thread 14406) (query_id: b60cfd44-8c83-497e-ba38-1c479869cd82) (query: DROP ROLE IF EXISTS role_c09ae1cc_70e9_11ef_88aa_6975d6534d58_unknown
) Received signal Segmentation fault (11)
2024.09.12 12:31:23.299220 [ 15136 ] {} <Fatal> BaseDaemon: Address: 0xffff9c544047. Access: . Attempted access has violated the permissions assigned to the memory area.
2024.09.12 12:31:23.299504 [ 15136 ] {} <Fatal> BaseDaemon: Stack trace: 0x000000000c558228 0x0000ffff9dffb7dc 0x000000000ee226a1 0x000000000ee1dbac 0x000000000ee21b78 0x000000000eda3fa8 0x000000000ed95bc4 0x000000000edff1e4 0x000000000ff172c8 0x000000000ff16c98 0x000000000fd8a6a0 0x000000000fd86a08 0x0000000010caef0c 0x0000000010cc5588 0x000000001344be98 0x000000001344c374 0x00000000134173b4 0x00000000134157f0 0x0000ffff9df89624 0x0000ffff9dee062c
2024.09.12 12:31:23.300171 [ 15136 ] {} <Fatal> BaseDaemon: 0. signalHandler(int, siginfo_t*, void*) @ 0x000000000c558228
2024.09.12 12:31:23.300411 [ 15136 ] {} <Fatal> BaseDaemon: 1. ? @ 0x0000ffff9dffb7dc
2024.09.12 12:31:23.300696 [ 15136 ] {} <Fatal> BaseDaemon: 2. std::__tree_iterator<std::__value_type<String, StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag>>, std::__tree_node<std::__value_type<String, StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag>>, void*>*, long> std::__tree<std::__value_type<String, StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag>>, std::__map_value_compare<String, std::__value_type<String, StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag>>, std::less<String>, true>, std::allocator<std::__value_type<String, StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag>>>>::find<String>(String const&) @ 0x000000000ee226a1
2024.09.12 12:31:23.301036 [ 15136 ] {} <Fatal> BaseDaemon: 3. DB::LDAPAccessStorage::applyRoleChangeNoLock(bool, StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag> const&, String const&) @ 0x000000000ee1dbac
2024.09.12 12:31:23.301585 [ 15136 ] {} <Fatal> BaseDaemon: 4. void std::__function::__policy_invoker<void (StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag> const&, std::shared_ptr<DB::IAccessEntity const> const&)>::__call_impl<std::__function::__default_alloc_func<DB::LDAPAccessStorage::setConfiguration(Poco::Util::AbstractConfiguration const&, String const&)::$_0, void (StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag> const&, std::shared_ptr<DB::IAccessEntity const> const&)>>(std::__function::__policy_storage const*, StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag> const&, std::shared_ptr<DB::IAccessEntity const> const&) @ 0x000000000ee21b78
2024.09.12 12:31:23.301863 [ 15136 ] {} <Fatal> BaseDaemon: 5. DB::AccessChangesNotifier::sendNotifications() @ 0x000000000eda3fa8
2024.09.12 12:31:23.302085 [ 15136 ] {} <Fatal> BaseDaemon: 6. DB::AccessControl::removeImpl(StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag> const&, bool) @ 0x000000000ed95bc4
2024.09.12 12:31:23.302361 [ 15136 ] {} <Fatal> BaseDaemon: 7. DB::IAccessStorage::remove(std::vector<StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag>, std::allocator<StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag>>> const&, bool) @ 0x000000000edff1e4
2024.09.12 12:31:23.302665 [ 15136 ] {} <Fatal> BaseDaemon: 8. DB::InterpreterDropAccessEntityQuery::execute()::$_0::operator()(std::vector<String, std::allocator<String>> const&, String const&) const @ 0x000000000ff172c8
2024.09.12 12:31:23.303040 [ 15136 ] {} <Fatal> BaseDaemon: 9. DB::InterpreterDropAccessEntityQuery::execute() @ 0x000000000ff16c98
2024.09.12 12:31:23.303357 [ 15136 ] {} <Fatal> BaseDaemon: 10. DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x000000000fd8a6a0
2024.09.12 12:31:23.303652 [ 15136 ] {} <Fatal> BaseDaemon: 11. DB::executeQuery(String const&, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum) @ 0x000000000fd86a08
2024.09.12 12:31:23.304005 [ 15136 ] {} <Fatal> BaseDaemon: 12. DB::TCPHandler::runImpl() @ 0x0000000010caef0c
2024.09.12 12:31:23.304282 [ 15136 ] {} <Fatal> BaseDaemon: 13. DB::TCPHandler::run() @ 0x0000000010cc5588
2024.09.12 12:31:23.304535 [ 15136 ] {} <Fatal> BaseDaemon: 14. Poco::Net::TCPServerConnection::start() @ 0x000000001344be98
2024.09.12 12:31:23.304793 [ 15136 ] {} <Fatal> BaseDaemon: 15. Poco::Net::TCPServerDispatcher::run() @ 0x000000001344c374
2024.09.12 12:31:23.304974 [ 15136 ] {} <Fatal> BaseDaemon: 16. Poco::PooledThread::run() @ 0x00000000134173b4
2024.09.12 12:31:23.305156 [ 15136 ] {} <Fatal> BaseDaemon: 17. Poco::ThreadImpl::runnableEntry(void*) @ 0x00000000134157f0
2024.09.12 12:31:23.305486 [ 15136 ] {} <Fatal> BaseDaemon: 18. start_thread @ 0x0000000000007624
2024.09.12 12:31:23.305919 [ 15136 ] {} <Fatal> BaseDaemon: 19. ? @ 0x00000000000d162c
2024.09.12 12:31:23.306105 [ 15136 ] {} <Fatal> BaseDaemon: Integrity check of the executable skipped because the reference checksum could not be read.
2024.09.12 12:31:23.306652 [ 15136 ] {} <Fatal> BaseDaemon: Report this error to https://github.com/ClickHouse/ClickHouse/issues
2024.09.12 12:31:23.307002 [ 15136 ] {} <Fatal> BaseDaemon: Changed settings: use_uncompressed_cache = false, load_balancing = 'random', max_memory_usage = 10000000000
```

### PR DESCRIPTION
**Pull request by zvonand**
Fix `heap-use-after-free` on DROP of LDAP-related roles (e.g. https://s3.amazonaws.com/clickhouse-test-reports/68355/80b28e33bb3d9fa3e1bbc2b1867086d8cc0a0a70/integration_tests__asan__old_analyzer__[4_6]//home/ubuntu/actions-runner/_work/_temp/test/output_dir/integration_run_parallel1_0.log)

Add missing `depends_on` in attempt to fix occasional fails of `test_ldap_external_user_directory`.

Closes #69555, maybe closes #65762

### Changelog category (leave one):
- Critical Bug Fix (crash, LOGICAL_ERROR, data loss, RBAC)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix crash on drop or rename a role that is used in LDAP external user directory

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit 1ee3a155be917497d17089a79407bccd66891a3d with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/68355/1ee3a155be917497d17089a79407bccd66891a3d/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Builds</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/8cdbe6333be2c2b720db1f7d0730a70ae33312c5/builds/report.html">✅ success</a></td></tr>
<tr><td>ClickBench</td><td>Runs [ClickBench](https://github.com/ClickHouse/ClickBench/) with instant-attach table</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/8cdbe6333be2c2b720db1f7d0730a70ae33312c5/clickbench__aarch64_.html">✅ success</a></td></tr>
<tr><td>Compatibility check</td><td>Checks that <b>clickhouse</b> binary runs on distributions with old libc versions. If it fails, ask a maintainer for help</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/8cdbe6333be2c2b720db1f7d0730a70ae33312c5/compatibility_check__aarch64_.html">✅ success</a></td></tr>
<tr><td>Docker keeper image</td><td>The check to build and optionally push the mentioned image to docker hub</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/8cdbe6333be2c2b720db1f7d0730a70ae33312c5/docker_keeper_image.html">✅ success</a></td></tr>
<tr><td>Docker server image</td><td>The check to build and optionally push the mentioned image to docker hub</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/8cdbe6333be2c2b720db1f7d0730a70ae33312c5/docker_server_image.html">✅ success</a></td></tr>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td>✅ success</td></tr>
<tr><td>Fast test</td><td>Normally this is the first check that is ran for a PR. It builds ClickHouse and runs most of <a href="https://clickhouse.com/docs/en/development/tests#functional-tests">stateless functional tests</a>, omitting some. If it fails, further checks are not started until it is fixed. Look at the report to see which tests fail, then reproduce the failure locally as described <a href="https://clickhouse.com/docs/en/development/tests#functional-test-locally">here</a></td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/8cdbe6333be2c2b720db1f7d0730a70ae33312c5/fast_test.html">✅ success</a></td></tr>
<tr><td>Flaky tests</td><td>Checks if new added or modified tests are flaky by running them repeatedly, in parallel, with more randomization. Functional tests are run 100 times with address sanitizer, and additional randomization of thread scheduling. Integration tests are run up to 10 times. If at least once a new test has failed, or was too long, this check will be red. We don't allow flaky tests, read <a href="https://clickhouse.com/blog/decorating-a-christmas-tree-with-the-help-of-flaky-tests/">the doc</a></td><td>✅ success</td></tr>
<tr><td>Install packages</td><td>Checks that the built packages are installable in a clear environment</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/8cdbe6333be2c2b720db1f7d0730a70ae33312c5/install_packages__aarch64_.html">✅ success</a></td></tr>
<tr><td>Performance Comparison</td><td>Measure changes in query performance. The performance test report is described in detail <a href="https://github.com/ClickHouse/ClickHouse/tree/master/docker/test/performance-comparison#how-to-read-the-report">here</a>. In square brackets are the optional part/total tests</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/8cdbe6333be2c2b720db1f7d0730a70ae33312c5/performance_comparison__release__[1_4]/report.html">✅ success</a></td></tr>
<tr><td>Stateful tests</td><td>Runs stateful functional tests for ClickHouse binaries built in various configurations -- release, debug, with sanitizers, etc</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/8cdbe6333be2c2b720db1f7d0730a70ae33312c5/stateful_tests__aarch64_.html">✅ success</a></td></tr>
<tr><td>Stateless tests</td><td>Runs stateless functional tests for ClickHouse binaries built in various configurations -- release, debug, with sanitizers, etc</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/8cdbe6333be2c2b720db1f7d0730a70ae33312c5/stateless_tests__aarch64_.html">✅ success</a></td></tr>
<tr><td>Stress test</td><td>Runs stateless functional tests concurrently from several clients to detect concurrency-related errors</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/8cdbe6333be2c2b720db1f7d0730a70ae33312c5/stress_test__debug_.html">✅ success</a></td></tr>
<tr><td>Unit tests</td><td>Runs the unit tests for different release types</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/8cdbe6333be2c2b720db1f7d0730a70ae33312c5/unit_tests__asan_.html">✅ success</a></td></tr>
<tr><td>Upgrade check</td><td>Runs stress tests on server version from last release and then tries to upgrade it to the version from the PR. It checks if the new server can successfully startup without any errors, crashes or sanitizer asserts</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/67541/8e6096dee72acc5ee75eb05ddfb9384767f52648/upgrade_check__debug_.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by zvonand** [comment]
`test_role_mapping` is also found to be flaky: 

```
    def test_role_mapping(ldap_cluster):
>       instance.query("CREATE ROLE role_1")

..................
E           helpers.client.QueryRuntimeException: Client failed! Return code: 237, stderr: Received exception from server (version 24.9.1):
E           Code: 493. DB::Exception: Received from 172.16.1.3:9000. DB::Exception: role `role_1`: cannot insert because role `role_1` already exists in local_directory. Stack trace:
E           
E           0. ./contrib/llvm-project/libcxx/include/exception:141: Poco::Exception::Exception(String const&, int) @ 0x00000000341ac174
E           1. ./build_docker/./src/Common/Exception.cpp:109: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000001ad36969
E           2. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x000000000aa52445
E           3. ./src/Common/Exception.h:128: DB::IAccessStorage::throwNameCollisionCannotInsert(DB::AccessEntityType, String const&) const @ 0x0000000026df6137
E           4. ./build_docker/./src/Access/DiskAccessStorage.cpp:536: DB::DiskAccessStorage::insertNoLock(StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag> const&, std::shared_ptr<DB::IAccessEntity const> const&, bool, bool, bool) @ 0x0000000026d85297
E           5. ./build_docker/./src/Access/DiskAccessStorage.cpp:513: DB::DiskAccessStorage::insertImpl(StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag> const&, std::shared_ptr<DB::IAccessEntity const> const&, bool, bool) @ 0x0000000026d87328
E           6. ./build_docker/./src/Access/MultipleAccessStorage.cpp:379: DB::MultipleAccessStorage::insertImpl(StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag> const&, std::shared_ptr<DB::IAccessEntity const> const&, bool, bool) @ 0x0000000026e66ea8
E           7. ./build_docker/./src/Access/AccessControl.cpp:549: DB::AccessControl::insertImpl(StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag> const&, std::shared_ptr<DB::IAccessEntity const> const&, bool, bool) @ 0x0000000026c97fcf
E           8. ./build_docker/./src/Access/IAccessStorage.cpp:195: DB::IAccessStorage::insert(std::vector<std::shared_ptr<DB::IAccessEntity const>, std::allocator<std::shared_ptr<DB::IAccessEntity const>>> const&, std::vector<StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag>, std::allocator<StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag>>> const&, bool, bool) @ 0x0000000026ded393
E           9. ./build_docker/./src/Access/IAccessStorage.cpp:201: DB::IAccessStorage::insert(std::vector<std::shared_ptr<DB::IAccessEntity const>, std::allocator<std::shared_ptr<DB::IAccessEntity const>>> const&, bool, bool) @ 0x0000000026ded10c
E           10. ./build_docker/./src/Interpreters/Access/InterpreterCreateRoleQuery.cpp:115: DB::InterpreterCreateRoleQuery::execute() @ 0x000000002a938727
E           11. ./build_docker/./src/Interpreters/executeQuery.cpp:1229: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x0000000029a5efe0
E           12. ./build_docker/./src/Interpreters/executeQuery.cpp:1395: DB::executeQuery(String const&, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum) @ 0x0000000029a58665
E           13. ./build_docker/./src/Server/TCPHandler.cpp:564: DB::TCPHandler::runImpl() @ 0x000000002d0362c5
E           14. ./build_docker/./src/Server/TCPHandler.cpp:2484: DB::TCPHandler::run() @ 0x000000002d068520
E           15. ./build_docker/./base/poco/Net/src/TCPServerConnection.cpp:43: Poco::Net::TCPServerConnection::start() @ 0x000000003438990f
E           16. ./build_docker/./base/poco/Net/src/TCPServerDispatcher.cpp:115: Poco::Net::TCPServerDispatcher::run() @ 0x000000003438a4f7
E           17. ./build_docker/./base/poco/Foundation/src/ThreadPool.cpp:205: Poco::PooledThread::run() @ 0x000000003428cc0b
E           18. ./base/poco/Foundation/src/Thread_POSIX.cpp:335: Poco::ThreadImpl::runnableEntry(void*) @ 0x0000000034286d68
E           19. asan_thread_start(void*) @ 0x000000000aa07059
E           20. ? @ 0x00007f789fde9ac3
E           21. ? @ 0x00007f789fe7b850
E           . (ACCESS_ENTITY_ALREADY_EXISTS)
E           (query: CREATE ROLE role_1)

```

**Comment by zvonand** [comment]
A sanitizer alert in the latest run, and no changes to CH code itself. Needs a closer look

File: src/Access/LDAPAccessStorage.cpp [hunk 1]
```diff
@@ -128,7 +128,7 @@ void LDAPAccessStorage::processRoleChange(const UUID & id, const AccessEntityPtr
     {
         if (it != granted_role_names.end()) // Removed a granted role.
         {
```
Comment (Algunenano) [inline src/Access/LDAPAccessStorage.cpp:130]: A comment with an explanation would be nice. The problem is that it->second is a reference to an object held by `granted_role_names`. Then in applyRoleChangeNoLock we first remove the object from `granted_role_names`, then try to remove the name from `granted_role_ids` but we are using the reference that has been just destroyed.

This problem also happens a few lines above with RENAME (L118), which should be adjusted too.

It seems this API is really error prone. Could we improve it instead?

Comment (Algunenano) [reply src/Access/LDAPAccessStorage.cpp:130]: I mean the function definition (`void LDAPAccessStorage::applyRoleChangeNoLock(bool grant, const UUID & role_id, const String & role_name)`) and how it's used to remove roles. You need to know exactly how things are destroyed and how to manage the lifetime of the iterator, which seems suboptimal.

**Pull request merged**